[Issue 203] Add report dashboard layout

dalogsdon opened the pull request:
Resolves #203

parente commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-metadata.js:
```diff
@@ -0,0 +1,219 @@
+                }
+            }
+        }
+**************/
```
:+1: for moving all the metadata business to its own module.

parente commented:
The "Hidden Cells" annotation in the UI seems to have disappeared in grid layout mode:

![screen shot 2016-04-04 at 2 39 50 pm](https://cloud.githubusercontent.com/assets/153745/14259050/23418610-fa73-11e5-89ff-10e3550470c7.png)

parente commented:
When taking an existing notebook and going from grid layout to report layout, the spacing between cell rows goes crazy. Try it on the scotch notebook.

Also: it seems mighty dangerous that a single misclick will potentially lose a complete dashboard layout when the user accidentally switches modes.

parente commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-actions.js:
```diff
@@ -2,131 +2,138 @@
-        scrollToBottom = IPython.Notebook.prototype.scroll_to_bottom;
-        IPython.Notebook.prototype.scroll_to_bottom = function() {};
+    function getFaIconClass(el) {
+        return $(el).attr('class').split(' ').filter(function(value) {
```
> dashboard-actions.js?v=20160404185313:55 Uncaught TypeError: Cannot read property 'split' of undefined

When opening the scotch-demo dashboard for the first time with these changes.

dalogsdon commented:
> The "Hidden Cells" annotation in the UI seems to have disappeared in grid layout mode

I was seeing the label during development. I did a fix this morning for a minor issue I saw, maybe that broke it. l'll have a look.



> When taking an existing notebook and going from grid layout to report layout, the spacing between cell rows goes crazy.

Not sure what you mean by goes crazy. When doing the transition myself, I didn't notice anything unusual. I'll try with scotch notebook.


> it seems mighty dangerous that a single misclick will potentially lose a complete dashboard layout when the user accidentally switches modes

That is why I put the layout selection in a dropdown menu. It is an attempt to make the action to switch layout more intentional. If you feel it is still too easy, we could experiment with a confirmation, though I was hoping to avoid any popup dialogs.

I've also just realized I put the toolbar layout selection in a dropdown but kept the menu items flat. I could also put the layout selection in a sub-menu to help mitigate any accidental layout wipeouts.

parente commented:
> Not sure what you mean by goes crazy. When doing the transition myself, I didn't notice anything unusual. I'll try with scotch notebook.

There's huge amounts of cell spacing between the cells when flipping the notebook between modes. A refresh clears things up. I bet the grid is not being computed correctly.

> hoping to avoid any popup dialogs.

Definitely do NOT want this. I'm more thinking about how do we set user expectations that once they switch, the layout is lost. They only way to find out that hard truth at the moment is to try it.

dalogsdon commented:
There is an issue with switching to report layout with pre-existing hidden cells. This seems to cause both the missing hidden label and the spacing issue.

dalogsdon commented:
An alternative to the lost layout issue is to retain all layout metadata, so the user can freely move between layouts as desired. Not sure if this would be confusing or not.

dalogsdon commented:
Ah, I put some debounced code to show the hidden area after the Gridstack is fully built, but that only happens if there are cells with no grid metadata in which case we add a widget which triggers the grid change event. If the notebook already has the layout metadata this code will not fire. I'll have to redo when the hidden area is recomputed to handle this case.

parente commented:
> An alternative to the lost layout issue is to retain all layout metadata, so the user can freely move between layouts as desired. Not sure if this would be confusing or not.

The problem with this is that the user would have a hard time knowing which is active. "The last one you viewed in layout mode" is a pretty weak indicator.

I think overwriting is fine for now. When we do the metadata proposal for https://github.com/jupyter/roadmap/pull/8, we can consider multiple named layouts.

dalogsdon commented:
Another hidden-cell related issue is that I was wiping out hidden data when switching to report layout. I've updated the metadata module to respect hidden cells when updating layout for all cells. 

@parente With the fixes I've done for hidden cells, the scotch demo looks correct now.

parente commented:
OK I'll give it another shot.

parente commented:
Notebook with no dashboard layout yet results in `Uncaught TypeError: Cannot read property 'dashboard' of undefined` from `:9500/nbextensions/jupyter_dashboards/notebook/dashboard-view/dashboard-metadata.js?v=20160405115407:109 `

parente commented:
Still getting bad spacing / layout in certain situations:

1. Create a new notebook
2. Enter and run x = 1, y = 2, x + y in 3 different cells.
3. Click grid layout button (not the dropdown since that results in the bug in my previous comment)
4. Grid layout looks OK.
5. Switch it to the report layout.

Hidden cells text disappears. The one lone cell is moved way down the page with too much whitespace around it. (Layout is fubar.)

![screen shot 2016-04-05 at 8 02 06 am](https://cloud.githubusercontent.com/assets/153745/14281058/c771f9ea-fb04-11e5-9adc-08af4c1896fd.png)

parente commented:
Marked WIP so that we know there's issues still being worked here.

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-actions.js:
```diff
@@ -2,131 +2,138 @@
-            .addClass('active')
-            .siblings().removeClass('active');
+    function updateAuthoringBtn(state) {
+        if (/^db-auth-/.test(state)) {
```
If we're setting up (basically) an ENUM at the beginning of the file (`var STATE = {...}`), I think we should use that everywhere, rather than sometimes also using the values. I know it will be more verbose, but it will make it easier to maintain and search in the future.

If you really need to map from values to enum, that should be refactored into its own getter function, preferably right below the definition of the enum itself.

(Here and elsewhere).

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-metadata.js:
```diff
@@ -0,0 +1,235 @@
+        /**
+         * @return {string} dashboard layout type (defaults to grid layout if not set in notebook metadata)
+         */
+        getDashboardLayout: function() {
```
We should start using the ES5 getter/setter syntax for these types of "properties" (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get).  Allows consumers to write code like `foo = metadata.dashboardLayout` to get and `metadata.dashboardLayout = foo` to set.

Also, getters and setters for same properties should be defined next to each other in this file.

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-view.js:
```diff
@@ -54,26 +54,53 @@ define([
 
     // dashboard-actions depends on requirejs text plugin
-    require(['./dashboard-actions'], function(DashboardActions) {
+    require(['./dashboard-actions', './dashboard-metadata'], function(DashboardActions, Metadata) {
```
Seems a bit odd to me for this module to use `dashboard-metadata`. I would think that should only be used by `dashboard.js`. Was there a specific reason to bring that up a level?

parente replied:
When we have to start supporting 2 metadata formats potentially, the old one + the new official Jupyter one, then it might be nice to have this separation.

dalogsdon replied:
My reasoning was that the distinction of 'grid' and 'report' layout would happen at a higher level than `dashboard.js`, which would be a Gridstack layout that can take parameters. The `dashboard-view.js` would modify the parameters to switch layouts between grid and report. We could just as easily bake these layouts into the `dashboard.js` module, but if other non-Gridstack layouts are added in the future they will have to be managed in other placed. I decided it would make the most sense if `dashboard-view.js` managed the layout by instantiating the proper layout modue with the proper parameters for the selected layout.

It may be good to do some module renaming to more accurately specify their roles in the dashboard extension:

`dashboard-view.js` --> `dashboard.js`
`dashboard.js` --> `gridstack-dashboard.js`

In the future there could be additional layout modules such as `slide-dashboard.js` or `multipage-dashboard.js`.

jhpedemonte replied:
OK, makes sense.

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard.js:
```diff
@@ -466,11 +383,19 @@ define([
+            $cell.attr('data-gs-width', width)
+                 .attr('data-gs-height', height);
+
+            if (layout.hasOwnProperty('col') && layout.hasOwnProperty('row')) {
```
Should just be able to do `if (layout.col && layout.row)`, no need for `hasOwnProperty`.

dalogsdon replied:
zero-based?

jhpedemonte replied:
Ohhhhh!

dalogsdon commented:
I merged in the grid move buttons and updated based on suggestions from @jhpedemonte.

@parente It looks like the metadata wasn't getting fully saved when switching between layouts. I've updated how the metadata is saved so your issue should be fixed.

dalogsdon commented:
I just noticed an issue related to the getter/setter updating I did earlier. It is trying set the layout before the metadata is initialized. I've updated to fix.

parente commented:
No major issues now that I can see when manually testing. One inconsistency that I think should get fixed: When I'm in report mode, the close X is in the top right of the cell:

<img width="955" alt="screen shot 2016-04-07 at 7 42 59 am" src="https://cloud.githubusercontent.com/assets/153745/14349976/7f5e883a-fc94-11e5-9e2e-e6e087ce5f50.png">

In grid mode, the X shifts to the left of the move buttons:

<img width="488" alt="screen shot 2016-04-07 at 7 42 51 am" src="https://cloud.githubusercontent.com/assets/153745/14349994/87824fe2-fc94-11e5-9266-951c7cae47c3.png">

I'm trained to look for close either in the top right or top left of a window, so I think it should stay put.

parente commented:
Testing on 2016-04-07:

----

One inconsistency that I think should get fixed: When I'm in report mode, the close X is in the top right of the cell:

<img width="955" alt="screen shot 2016-04-07 at 7 42 59 am" src="https://cloud.githubusercontent.com/assets/153745/14349976/7f5e883a-fc94-11e5-9e2e-e6e087ce5f50.png">

In grid mode, the X shifts to the left of the move buttons:

<img width="488" alt="screen shot 2016-04-07 at 7 42 51 am" src="https://cloud.githubusercontent.com/assets/153745/14349994/87824fe2-fc94-11e5-9266-951c7cae47c3.png">

I'm trained to look for close either in the top right or top left of a window, so I think it should stay put.

----

Spacing for a 1 line markdown cell is huge in the test_layout_basic dashboard when going from grid to report. Looks like height is not recomputed properly.

![screen shot 2016-04-07 at 7 47 01 am](https://cloud.githubusercontent.com/assets/153745/14350043/f019d41c-fc94-11e5-9175-44a2f84856ff.png)

----

Test layout has a similar problem. Plus, the first cell is too short and clips the markdown text.

![screen shot 2016-04-07 at 7 48 25 am](https://cloud.githubusercontent.com/assets/153745/14350086/30c24d46-fc95-11e5-8acb-92a89c97e819.png)


----

Occurs to me that we also need to update our old Thebe-based "Local deploy" template logic to support report mode. It may "just work" like it says in https://github.com/jupyter-incubator/dashboards_server/issues/157, but I've opened https://github.com/jupyter-incubator/dashboards_bundlers/issues/25 to for testing to make sure it does.

dalogsdon commented:
> We could enable vertical sizing for cells in report mode

It should be enabled. I tried it just now and it worked. Perhaps the cell content was blocking the resize handle?

> but then I think we've defeated the purpose of report mode entirely which is supposed to be as simple as "just let me pick which cells to show / hide"

Yes, that makes sense. We could override the CSS or create a new very simple layout module for report.

> One inconsistency that I think should get fixed: When I'm in report mode, the close X is in the top right of the cell

Good catch. This was fixed in #208 but then regressed when merging with #203 since the grid controls template was moved from inline string to a file. I can fix that.

I'll debug into the other issues you mentioned.

dalogsdon commented:
> Spacing for a 1 line markdown cell is huge in the test_layout_basic dashboard when going from grid to report. Looks like height is not recomputed properly.

That is because the height was already set in the grid view. No recomputation happens in that case. The suggestion to use auto sizing should fix this.

parente commented:
Thoughts on the latest:

* Yes. Report mode this way is better than the old report-mode-on-gridstack way. Good work.
* I'm still concerned about accidentally clicking the wrong layout in the menu or the drop down wiping out someone's layout in either mode. I still don't have a perfect idea of how to avoid it. Definitely don't want a confirmation dialog. Definitely don't want to retain both layouts permanently in the notebook file and have to have an "active" one. What about retaining both sets of metadata until next save, assuming we don't save immediately on layout switch? That way, the user can say "oh crap" and "undo" by switching back to the other layout mode with the layout info retained.
* It's a bit hard to get a sense of the true report layout because of how much space the cells take up. I know this is what preview is for. I think I'm just used to seeing the grid layout without hidden cells cluttering it, and that I'll get used to report mode being different.
* One widget/output overlap nit in report mode on a hidden cell in the scotch demo:

<img width="1027" alt="screen shot 2016-04-12 at 10 19 49 pm" src="https://cloud.githubusercontent.com/assets/153745/14481118/dbe0e2f2-00fd-11e6-8fc5-4bcb21384440.png">

As soon as I unhid the cell, the two stopped overlapping.

I asked @jtyberg to try this too.

jtyberg commented:
Report layout is a welcome addition for sure.  Here's my gut reaction after playing with it a bit:

* I have not yet formed an opinion on even _using_ two different layouts, e.g., grid and report.  The truth is, I can't think of many times where I've wanted both.
* My mind is agitated by the cells that say, "Hidden", but I can still see them.  It reminds me of when I was a kid and friends would say, "You can't see me.", but in fact, I could.  I want "hidden" cells to all but disappear from the report flow.  Perhaps they transform into a thin line (5 px height)?  Maybe when I hover on the line, it expands to full height, even though it is still "hidden".  Or maybe we make use of all that margin?  Show a thumbnail of the hidden cells on the right margin.  Click the thumbnail to unhide it.  Click the 'X' to shove it to the side again.  Or perhaps the hidden cells actually do hide when I move my mouse to the margin?  I'm just spit-balling here.
* I'll be honest, I'm not a big user of the Dashboard Preview mode.  In grid layout, I was happy just to move my mouse to the margin and let the white background fill in the cells.  That was good enough for me.  So I don't like the idea of having to switch over to Dashboard Preview just to see the Hidden cells actually hide.

dalogsdon commented:
> So I don't like the idea of having to switch over to Dashboard Preview just to see the Hidden cells actually hide

I agree that it is nice to be able to see the layout the same when authoring vs previewing.

> Show a thumbnail of the hidden cells on the right margin

This could work, perhaps on hover the hidden cell can be overlayed without affecting the dashboard flow, and clicking the thumbnail will actually add it back into the flow.

parente commented:
> Show a thumbnail of the hidden cells on the right margin

Let's be careful not to make it too complicated. The more complex the feature, the more bugs across browsers!

Maybe try some simple things first like more transparency, shorter height?

jhpedemonte commented:
Showing the page the same as in preview when not hovering, then expanding the hidden cells (and keeping them greyed out) on hover might be a good solution. Not terribly complex.

dalogsdon commented:
![dashboard-report-hover-1](https://cloud.githubusercontent.com/assets/6362426/14500866/647a7892-0169-11e6-9a9a-eab8fac45bec.gif)

parente commented:
What's it look like when it's a long notebook, you're scrolled, and you hover / unhover? Do you lose your spot?

dalogsdon commented:
Yeah in scotch dashboard things get pushed around quite a bit in hover vs unhover. I think this will require a JS solution to maintain good usability. Minimally, I can play around with the container scroll position to see if expand on hover will work well.

parente commented:
Are we quickly coming back to having the hidden cells out of the flow at the bottom? 😱

jtyberg commented:
Would it be difficult to combine both the disappearing off-hover trick WITH shrinking the "hidden" cells when hovering?  And when you hover over a particular hidden cell, only then expand that cell to full height? Perhaps not so much motion that way.

dalogsdon commented:
Discussed with @jhpedemonte. We came up with a non-hover based solution that should work better. Will update when that is ready to test.

dalogsdon commented:
Updated to manually collapse/expand adjacent hidden cells by clicking inline collapse buttons. They start as collapsed when first opening report layout:

![image](https://cloud.githubusercontent.com/assets/6362426/14512312/9364e810-01a3-11e6-81bf-1657fe6e92c9.png)

@parente @jtyberg Give it a go. Hopefully the collapsed cells can help give the impression of the report while in authoring view so you don't have to go to preview to see it. If it is still too intrusive we can experiment with reducing the collapsed size further. I've shied away from using the margins so far so things work the same in "mobile" width, but if needed we can start to experiment with that as well.

I've realized that this PR now contains more additions and deletions than the sum of all my previous commits. :open_mouth:

parente commented:
> I've realized that this PR now contains more additions and deletions than the sum of all my previous commits.

You're going to win! 🏆

parente commented:
I gotta do a bit more testing with the decl widgets to see how they behave in the DOM flow, but as far as the collapse/expand UX goes, count me in. I can definitely get a sense of what the entire thing will look like by collapsing the contiguous cells. And it's not too hard to expand, find a cell, add it, and re-collapse when I move through the notebook.

I tested a bunch of edge cases too like hiding a set of cells, collapsing them as a group, hiding another set of cells with a shown cell between the groups, and then hiding that one shown cell to make sure the two collapse groups join. They do. Nice work.

jtyberg commented:
Oh, yes.  This works for me.

I only have one comment:

* When I unhide a cell, I want to see it full height.  Currently it shows me a fixed height.

![dashboard_issue_211](https://cloud.githubusercontent.com/assets/359081/14535903/09ce68c4-023e-11e6-887d-b3a0a5c6e7f2.gif)

dalogsdon commented:
> When I unhide a cell, I want to see it full height. Currently it shows me a fixed height.

Good catch. That is a styling bug that happens when switching to the report layout before viewing. I've updated to fix.

dalogsdon commented:
I found one other issue. We have several HTML template files we add to the page. These have the Jupyter copyright headers in them, so we were adding copyright header comments to the document for each template. This PR made this issue more visible, so I've included a simple fix here, which is a requirejs plugin called "template" that will read a template file and remove any comment nodes.

parente commented:
It's also reasonable to just leave the Jupyter copyright out of the templates.

parente commented:
I tested with declarative widgets 0.4.5 by opening the streaming meetup demo, starting it running, then manipulating the cells visible in the report layout. No issues throughout. 

I say :shipit:

One final note: I'm on a desktop w/ monitor today and noticed it was a bit harder to quickly make out the difference between hidden and shown cells when not collapsed. I don't want to hold this up any longer, but let's keep it in mind as something to tweak maybe if others report it.

parente commented:
Make sure that if an older notebook has a grid layout with declaring as a grid layout, that it still shows properly when immediately opened in preview mode for backward compat.

dalogsdon commented:
`dashboard-metadata.js` will return `grid` when accessing `dashboardLayout` by default. This will cause a notebook that has grid metadata without declaring grid layout to render properly and will cause a default grid layout to be set if the notebook has no dashboard metadata. However, the layout will not be set in either of these cases, so these notebooks will still rely on the default layout being grid. I could update this so preview mode sets the layout to grid if not set.

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-actions.js:
```diff
@@ -2,131 +2,137 @@
+
+    function updateUIState(state) {
+        updateUrlState(state === STATE.DASHBOARD_PREVIEW);
+        setHeaderVisibility(state !== STATE.DASHBOARD_PREVIEW);
```
Pull out these equality checks into their own variable? May make it easier to read this function and avoid doing same call multiple times:

    var inDashboardPreview = state === STATE.DASHBOARD_PREVIEW;

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/layout/grid/layout.js:
```diff
@@ -600,24 +468,29 @@ define([
+    // computes metadata for a cell that doesn't have grid layout metadata
+    GridLayout.prototype._addNotebookCellToDashboard = function($cell) {
+        var dim = this._computeCellDimensions($cell);
+        return dim.isEmpty ? Promise.resolve() : this._showCell($cell);
```
Should that promise resolve with `$cell` in order to match promise returned from `_showCell`?

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/layout/grid/layout.js:
```diff
@@ -627,9 +500,22 @@ define([
+                    self.gridstack.container.off('change', onChange);
+                }
+            };
+            this.gridstack.container.on('change', onChange);
```
Namespace the `'change'` event, since you're calling `off` later.

dalogsdon replied:
`off` is getting passed the callback function, so namespacing is not necessary. Additionally, this function gets called many times in certain situations, so namespacing would get complex.

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/layout/grid/layout.js:
```diff
@@ -627,9 +500,22 @@ define([
+            // add widget to Gridstack, reject if it takes too long
+            var rejectTimeout = window.setTimeout(function() {
+                cellAddedPromise.reject('Timeout waiting for Gridstack to add widget', $cell);
+            }, 10000);
```
Please add a comment here as to why we need this reject timeout.

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/layout/grid/layout.js:
```diff
@@ -627,9 +500,22 @@ define([
+                cellAddedPromise.reject('Timeout waiting for Gridstack to add widget', $cell);
+            }, 10000);
+            var onChange = function(event, elements) {
+                if (elements.some(function(el) { return el.el.is($cell); })) {
```
You might just be able to do `if ($(elements).is($cell))`. If not, this should be refactored into its own function to make this bit more legible.

dalogsdon replied:
`elements` is an array of objects, each of which contains an `el` jQuery object; notice the contents of the `some` function is `el.el.is()`

jhpedemonte commented on jupyter_dashboards/nbextension/notebook/dashboard-view/layout/report/layout.js:
```diff
@@ -0,0 +1,169 @@
+
+    var cssLoaded = false;
+    var $CONTAINER = $('#notebook-container');
+    var SCROLL_CONTAINER = $('#site').get(0);
```
These will be the same across all layouts, so they should probably be passed in as options to the constructors.

dalogsdon commented:
Updated based on comments by @jhpedemonte

parente commented:
As soon as I open any notebook now, the header and toolbars are immediately hidden.

parente commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-actions.js:
```diff
@@ -2,131 +2,139 @@
+    function updateUIState(state) {
+        var isDashboardPreview = state === STATE.DASHBOARD_PREVIEW;
+        updateUrlState(isDashboardPreview);
+        setHeaderVisibility(isDashboardPreview);
```
Parameter needs to be negated.

parente commented on jupyter_dashboards/nbextension/notebook/dashboard-view/dashboard-actions.js:
```diff
@@ -2,131 +2,139 @@
-        $('#header-container').toggle(doShow && isHeaderVisible);
-        $('.header-bar').toggle(doShow && isHeaderVisible);
+        // hide or revert back to previous state
+        $('#header-container, .header-bar').toggle(doShow && isHeaderVisible);
```
Logic fails when loading the page with `?dashboard` and then trying to switch to notebook. `doShow=true && isHeaderVisible=false &rarr; false` so the toggle doesn't happen.

parente replied:
Update: logic fails in this situation because the values are first computed by the query string method and stored as true and then immediately again when add the toolbar is added and stored as false because they're already hidden.

parente commented:
I'll submit a PR to fix the toolbar state as a follow-on. This one has gone on long enough.
